File: todoforai_edge/fuse/path_mapping.py
"""Path ↔ URI mapping for the FUSE filesystem.

Port of fs/fuse/src/index.ts lines 32-56.
"""

from typing import Optional

SCHEMES = ["todoforai", "http", "https", "gdrive", "edge"]


def path_to_uri(path: str) -> Optional[str]:
    """Convert a FUSE path to a resource URI.

    /todoforai/abc123            → todoforai://attachment/abc123
    /http/example.com/file.txt   → http://example.com/file.txt
    /https/example.com/file.txt  → https://example.com/file.txt
    """
    parts = [p for p in path.split("/") if p]
    if not parts:
        return None

    scheme = parts[0]
    if scheme not in SCHEMES:
        return None

    if scheme == "todoforai":
        return f"todoforai://attachment/{'/'.join(parts[1:])}"
    return f"{scheme}://{'/'.join(parts[1:])}"


def is_scheme_root(path: str) -> bool:
    parts = [p for p in path.split("/") if p]
    return len(parts) == 1 and parts[0] in SCHEMES


def is_root(path: str) -> bool:
    return path == "/"

File: bun/src/fuse/path-mapping.ts
/**
 * Path ↔ URI mapping for the FUSE filesystem.
 * Port of todoforai_edge/fuse/path_mapping.py
 */

export const SCHEMES = ["todoforai", "http", "https", "gdrive", "edge"] as const;
const SCHEME_SET = new Set<string>(SCHEMES);

export function pathToUri(fsPath: string): string | null {
  const parts = fsPath.split("/").filter(Boolean);
  if (!parts.length) return null;

  const scheme = parts[0];
  if (!SCHEME_SET.has(scheme)) return null;

  if (scheme === "todoforai") {
    return `todoforai://attachment/${parts.slice(1).join("/")}`;
  }
  return `${scheme}://${parts.slice(1).join("/")}`;
}

export function isSchemeRoot(fsPath: string): boolean {
  const parts = fsPath.split("/").filter(Boolean);
  return parts.length === 1 && SCHEME_SET.has(parts[0]);
}

export function isRoot(fsPath: string): boolean {
  return fsPath === "/";
}
